### PR TITLE
Simplified leaky bucket

### PR DESCRIPTION
**Pull request by ilikeorangutans**
Realized that the leaky bucket could be rewritten without locks by harnessing the power of channels! :zap: 
@csaunders @chrisbutcher

### CODE REVIEW

File: bucket/leaky_bucket.go [hunk 1]
```diff
@@ -40,61 +49,37 @@ func (b *LeakyBucket) StopDripping() {
 	}()
 }
 
-func (b *LeakyBucket) UnsafeAvailable() int {
-	return b.available
+func (b *LeakyBucket) Available() int {
+	return len(b.bucket)
```
Comment (csaunders) [inline bucket/leaky_bucket.go]: I thought you couldn't read the length of items queued into a channel

File: bucket/leaky_bucket_test.go [hunk 2]
```diff
@@ -49,12 +51,21 @@ func TestGrabbingADropFromTheBucket(t *testing.T) {
...
 
 	assert.Equal(t, true, bucket.IsFull())
 	bucket.AddDrops()
 	assert.Equal(t, true, bucket.IsFull())
-	assert.Equal(t, 2, bucket.UnsafeAvailable())
+	assert.Equal(t, 2, bucket.Available())
+}
+
+func BenchmarkBucket(b *testing.B) {
+	config := Configuration{Size: 2, Refill: 1, Duration: time.Duration(1) * time.Millisecond}
```
Comment (csaunders) [inline bucket/leaky_bucket_test.go]: Would've been cool to have this test before and after to see if this new implementation performed better.

But, this code is way cleaner so I'm kinda yoloswag about that.

**Comment by csaunders** [comment]
@ilikeorangutans you are such a cool cat. Thanks for making this suck less :D

File: bucket/leaky_bucket.go (hunk 1 above)
Comment (ilikeorangutans) [inline bucket/leaky_bucket.go]: https://golang.org/pkg/builtin/#len :
> The len built-in function returns the length of v, according to its type:
> Channel: the number of elements queued (unread) in the channel buffer;

:)

File: bucket/leaky_bucket.go [hunk 3]
```diff
@@ -40,61 +49,37 @@ func (b *LeakyBucket) StopDripping() {
...
-		amount = b.refill
-	}
+func (b *LeakyBucket) fill(amount int) {
 	for i := 0; i < amount; i++ {
-		b.bucket <- true
-		b.locked(func() interface{} {
-			b.available++
-			return nil
-		})
+		select {
```
Comment (ilikeorangutans) [inline bucket/leaky_bucket.go]: This is really cool and what makes this work without locks. You can use selects not only on reading from channels but also for writing to them!

File: bucket/leaky_bucket.go (hunk 1 above)
Comment (csaunders) [inline bucket/leaky_bucket.go]: Can you tell I don't always read all the docs for a language I'm learning?
:P

On Fri, Dec 18, 2015 at 10:41 AM, Jakob Külzer <notifications@github.com>
wrote:

> In bucket/leaky_bucket.go
> <https://github.com/Shopify/themekit/pull/127#discussion_r48037085>:
>
> > @@ -40,61 +49,37 @@ func (b *LeakyBucket) StopDripping() {
> >  	}()
> >  }
> >
> > -func (b *LeakyBucket) UnsafeAvailable() int {
> > -	return b.available
> > +func (b *LeakyBucket) Available() int {
> > +	return len(b.bucket)
>
> https://golang.org/pkg/builtin/#len :
>
> The len built-in function returns the length of v, according to its type:
> Channel: the number of elements queued (unread) in the channel buffer;
>
> :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Shopify/themekit/pull/127/files#r48037085>.
>



-- 
Chris Saunders // Developer @ Shopify

File: bucket/leaky_bucket_test.go (hunk 2 above)
Comment (ilikeorangutans) [inline bucket/leaky_bucket_test.go]: Meh, I thought about benchmarking old vs new, but it's by no means a performance critical part. I originally added this so i could run the race detector, no to so much to test how fast it is.

File: bucket/leaky_bucket.go (hunk 1 above)
Comment (ilikeorangutans) [inline bucket/leaky_bucket.go]: To be honest, I had to read that up too because I wasn't sure! :)

File: bucket/leaky_bucket_test.go (hunk 2 above)
Comment (csaunders) [inline bucket/leaky_bucket_test.go]: That's a legit reason to do it!

On Fri, Dec 18, 2015 at 10:45 AM, Jakob Külzer <notifications@github.com>
wrote:

> In bucket/leaky_bucket_test.go
> <https://github.com/Shopify/themekit/pull/127#discussion_r48037529>:
>
> >  	bucket.TopUp()
> >
> >  	assert.Equal(t, true, bucket.IsFull())
> >  	bucket.AddDrops()
> >  	assert.Equal(t, true, bucket.IsFull())
> > -	assert.Equal(t, 2, bucket.UnsafeAvailable())
> > +	assert.Equal(t, 2, bucket.Available())
> > +}
> > +
> > +func BenchmarkBucket(b *testing.B) {
> > +	config := Configuration{Size: 2, Refill: 1, Duration: time.Duration(1) * time.Millisecond}
>
> Meh, I thought about benchmarking old vs new, but it's by no means a
> performance critical part. I originally added this so i could run the race
> detector, no to so much to test how fast it is.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Shopify/themekit/pull/127/files#r48037529>.
>



-- 
Chris Saunders // Developer @ Shopify

File: bucket/leaky_bucket_test.go [hunk 4]
```diff
@@ -9,27 +9,29 @@ import (
 func TestFillingABucket(t *testing.T) {
```
Comment (chrisbutcher) [inline bucket/leaky_bucket_test.go]: Is it worth making any assertions on `Available()` in this test case?

File: bucket/leaky_bucket_test.go [hunk 5]
```diff
@@ -9,27 +9,29 @@ import (
...
 	bucket.AddDrops()
-	assert.Equal(t, false, bucket.IsEmpty(), "The bucket should have 1 drop in it but has %d instead", bucket.UnsafeAvailable())
+	assert.Equal(t, false, bucket.IsEmpty(), "The bucket should have 1 drop in it but has %d instead", bucket.Available())
 	assert.Equal(t, false, bucket.IsFull(), "The bucket should not be full yet")
 	bucket.AddDrops()
-	assert.Equal(t, true, bucket.IsFull(), "The bucket should have 2 drops and be full but instead it has %d drops", bucket.UnsafeAvailable())
+	assert.Equal(t, true, bucket.IsFull(), "The bucket should have 2 drops and be full but instead it has %d drops", bucket.Available())
 }
 
 func TestToppingUpABucket(t *testing.T) {
```
Comment (chrisbutcher) [inline bucket/leaky_bucket_test.go]: Is it worth making any assertions on `IsEmpty()` in this test case?

File: bucket/leaky_bucket.go (hunk 3 above)
Comment (chrisbutcher) [inline bucket/leaky_bucket.go]: :sparkles: :heart:

**Comment by chrisbutcher** [comment]
LGTM

File: bucket/leaky_bucket_test.go (hunk 4 above)
Comment (ilikeorangutans) [inline bucket/leaky_bucket_test.go]: Probably doesn't hurt. Will add.

File: bucket/leaky_bucket_test.go (hunk 5 above)
Comment (ilikeorangutans) [inline bucket/leaky_bucket_test.go]: Will add.

**Pull request merged**